### DISCUSSION TITLE
Why doesn't this result in an infinite loop?

### DISCUSSION THREAD

**Discussion by Sawtaytoes** [ReactiveX/rxjs#6071] [Q&A] [answered]
Can someone explain why this isn't an infinite loop? Seems to be something about execution order with `startWith`. Might even be a bug; not sure.

This code has a subject calling `.next` on itself:
```js
const subject$ = new Subject();

subject$
.pipe(
  startWith(null),
  switchMap(() => (
    throwError('error')
    .pipe(
      catchError(() => {
        subject$
        .next()

        return of(null);
      }),
    )
  )),
)
.subscribe()
```

There are three fixes to make this infinitely loop:
1. Add a `delay` after `startWith`:
```js
startWith(null),
delay(0),
```
2. Wrap `subject$` in `merge` and include `of`:
> _**NOTE:** This fix doesn't work with `EMPTY` or `NEVER` probably because of the same issue as `startWith`._
```js
merge(
  of(null),
  subject$,
)
```
3. Add `setTimeout` before calling `.next` the first time through:
```js
catchError(() => {
  if (isFirstTime) {
    isFirstTime = false

    setTimeout(() => {
      subject$
      .next()
    })
  }
  else {
    subject$
    .next()
  }

  return of(null);
}),
```

Looks like this might be related to an issue I had a few years back when trying to do the same thing; calling a subject's `.next` from in its pipeline: https://github.com/ReactiveX/rxjs/issues/5410.

**Comment by cartant** [comment, marked as answer]
My guess is that `startWith` emits before it subscribes to its source, so it's not subscribed at the time the `next` call is made. Feature. Not a bug.

**Reply by Sawtaytoes** [reply]
Super weird to me. I thought it would first subscribe, then run pipeline operators. How does it run a pipeline operator before a subscription occurs?

Is there a way to `startWith` and `delay(0)` without affecting other parts of the pipeline? For instance, if I add `delay(0)`, it will delay all values. But if I only delay the `startWith`, that would allow my infinite loop to activate.

Can this be added to the docs for `startWith`? It implies it executes before values come in from the source, but not that the source isn't even subscribed at the point `startWith` executes.

What else executes in the pipeline before the source subscribes?

**Reply by kwonoj** [reply]
startWith is basically `concat` (https://github.com/ReactiveX/rxjs/blob/master/src/internal/operators/startWith.ts#L83), so no affect to nature of observable. however `delay` involves scheduler (https://github.com/ReactiveX/rxjs/blob/master/src/internal/operators/delay.ts#L55) changes synchronous observable to async.

You need to ensure what kind of observable to treat if you're trying to make reentrant. RxJS by design, does not gaurantee reentrant behavior and it's up to callers.

**Reply by Sawtaytoes** [reply]
Is there a way to achieve synchronous execution after the source subscribes using a scheduler?

I've never used those before, so I'm not sure if they're a solution to this problem, but looking at that source code, and `startWith` takes a scheduler.

**Reply by cartant** [reply]
Try:
```ts
.pipe(
  startWith(null),
  observeOn(queueScheduler),
```

**Reply by Sawtaytoes** [reply]
I used RxViz to test. Might not have the latest RxJS so who knows.

This is what it output, same as I had before; only the first value:

![image](https://user-images.githubusercontent.com/3948069/109742199-45b81400-7b94-11eb-8d40-ca25d8777767.png)

**Reply by cartant** [reply]
> same as I had before; only the first value

That's correct. I should have given it some more thought. `observeOn(queueScheduler)` is going to be ineffectual. The `queueScheduler` works by queueing any actions that are scheduled whilst it's running an already-scheduled action. You could, however, use it to ensure the `subject.next` call is made after the entirety of the `startWith` implementation like this:

```
const subject = new Subject<void>();
subject
  .pipe(
    (source) =>
      new Observable((subscriber) => {
        const subscription = queueScheduler.schedule(() => {
          source.pipe(startWith(null)).subscribe(subscriber);
        });
        return subscription;
      }),
    switchMap(() =>
      throwError("error").pipe(
        catchError(() => {
          queueScheduler.schedule(() => {
            subject.next();
          });
          return of(null);
        })
      )
    )
  )
  .subscribe();
```

Not that I am recommending doing such a thing.

**Reply by Sawtaytoes** [reply]
I wonder though. Without a hidden `setTimeout`, this solves my problem and ensures the code runs as expected.

The observable in `switchMap` could execute synchronously or asynchronously and `subject$.next` will do what I'd expect.

I'd name that operator something like `delayedStartWith`. It's super niche, but is important for this particular case where I have a subject doing a `.next` on itself but also needs to immediately execute a value on subscribe.

---

What are the problems doing it this way?

**Reply by kwonoj** [reply]
All time recommendations from RxJS is don't reentrant. If you already aware problems and willing to deal with it, you could choose escape hatches for your purpose. Core api probably won't able to consider it to expose as public interface.

For escape hatches, there maybe problems / or not - but it depends on your reentrant usecase since there's no single approach applicable for all reentrant cases.

**Comment by backbone87** [comment]
are we just talking about behavior or is there a use case to this?

**Reply by cartant** [reply]
> is there a use case to this?

IMO, no.
